language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/UsedTypeParams.scala

```diff
@@ -26,10 +31,191 @@ object UsedTypeParams {
     case iface.TypeCon(_, xs) => xs.toSet.flatMap(collectTypeParams)
     case iface.TypeNumeric(_) => Set.empty
   }
+
+  final case class LookupType[+RF, +VF](
+      run: iface.TypeConName => Option[(iface.DefDataType[RF, VF], LookupType[RF, VF])]
+  )
+
+  import VarianceConstraint.BaseResolution
+  import ScopedDataType.{Name => I}
+  import Variance._
+  private[this] type TVar = Ref.Name
+
+  final class ResolvedVariance private (private val prior: Map[I, ImmArraySeq[Variance]]) {
+    def allCovariantVars(
+        dt: I,
+        ei: iface.EnvironmentInterface,
+    ): (ResolvedVariance, ImmArraySeq[Variance]) = {
+      val resolved = covariantVars(dt, (i: I) => ei.typeDecls get i map (_.`type`))
+      (
+        resolved,
+        resolved.prior.getOrElse(
+          dt,
+          ei.typeDecls.getOrElse(dt, sys.error(s"$dt not found")).`type`.typeVars map (_ =>
+            Covariant
+          ),
+        ),
+      )
+    }
+
+    /** Variance of `sdt.typeVars` in order. */
+    private[this] def covariantVars[RF <: iface.Type, VF <: iface.Type](
+        dt: I,
+        lookupType: I => Option[iface.DefDataType[RF, VF]],
+    ): ResolvedVariance = {
+      import iface._
+
+      def lookupOrFail(i: I) = lookupType(i) getOrElse sys.error(s"$i not found")
+
+      def goType(dt: I, seen: Set[I], contexts: Map[I, Set[TVar]])(
+          typ: iface.Type
+      ): VarianceConstraint = typ match {
+        case TypeVar(name) =>
+          // while we default to Covariant at a later step,
+          // absence *does not* mean set-to-Covariant at this step
+          VarianceConstraint(
+            Map(dt -> Map(name -> Covariant)),
+            DelayedResolution(
+              if (contexts.nonEmpty) Map(dt -> Map(name -> contexts)) else Map.empty
+            ),
+          )
+
+        case TypePrim(pt, typArgs) =>
+          import PrimType.{Map => _, _}
+          pt match {
+            case GenMap =>
+              val Seq(kt, vt) = typArgs
+              // we don't need to inspect `kt` any further than enumerating it;
+              // every occurrence therein is invariant
+              goType(dt, seen, contexts)(vt) |+| VarianceConstraint.base(
+                Map(dt -> collectTypeParams(kt).view.map((_, Invariant)).toMap)
+              )
+            case Bool | Int64 | Text | Date | Timestamp | Party | ContractId | List | Unit |
+                Optional | TextMap =>
+              // this is only safe for all-params-covariant cases
+              typArgs foldMap goType(dt, seen, contexts)
+          }
+
+        case TypeCon(TypeConName(tcName), typArgs) =>
+          val refDdt = lookupOrFail(tcName)
+          val argConstraints = refDdt.typeVars zip typArgs foldMap { case (tv, ta) =>
+            goType(dt, seen, contexts |+| Map(tcName -> Set(tv)))(ta)
+          }
+          val refConstraints =
+            if (seen(tcName)) mzero[VarianceConstraint] else goSdt(tcName, seen + tcName)(refDdt)
```

Comment:
I don’t understand how the `seen` stuff works. This doesn’t actually stop you from calling `goSdt` multiple times on a type, e.g., you could call `goType` once per type arg in line `102` since you don’t update `seen` across calls to `goType`. Is that intentional, if so why?

Reply:
`seen` _only_ prevents cycles in `goSdt`.